text/0000-stack-buildpacks.md

```diff
@@ -0,0 +1,606 @@
+
+If a stack buildpack provides a mixin that is not required, the stack buildpack MAY pass detection. This is in contrast to a userspace buildpack providing a dependency that is not required, which fails detection. If a mixin is required for a [single stage only](https://github.com/buildpacks/rfcs/pull/109) with the `build:` or `run:` prefix, a stack buildpack may provide it for both stages without failing detection. However, it will not be included in the Buildpack Build Plan during the stage where it is not required.
+
+During the build phase, the lifecycle will create a build plan containing only the entries required during that stage (build or run).
+* If a mixin is required for "run" stage only, it will not appear in the build plan entries during build
+* If a mixin is required for "build" stage only, it will not appear in the build plan entries during extend
+
+The entries of the build plan during the build and extend phases will have a new `mixin` key. For example:
+
+```
+[[entries]]
+name = "libpq"
+mixin = true
+```
+
+A Stack Buildpack that needs to install mixins must select them from the build plan.
+
+## Caching and Restoring
+
+During the export phase, the lifecycle will store any snapshot layers created during the build phase in the cache.
+
+During the restore phase of the next build, the lifecycle will download the snapshot layer with the cache and store it as a tarball in the `<layers>` directory (i.e. it will not extract it). The `restorer` cannot extract the snapshot because it will not run with root privileges. In addition, the `restorer` may run in a different container than the build, which means changes made to the base image are not guaranteed to carry forward. The `builder` will run with root privileges, but will drop privileges after executing stackpacks and before running userspace buildpacks.
+
+During the build phase, the lifecycle will extract and apply each snapshot tarball before running stack buildpacks, but after a snapshot-baseline has been captured. This ensures that all changes from the previous snapshot are preserved even if the stack buildpack does not make any additional changes.
+
+## Rebasing an App
+
+Before a launch image is rebased, the platform must re-run the any stackpacks that were used to build the launch image against the new run-image. The image containing the stack buildpacks and the builder binary must be provided to the rebaser operation as an argument. A platform may choose to provide the same stack buildpacks and builder binary used during the build that create the launch-image being rebased, or it may provide updates versions (which may increase the risk of something failing in the rebase process).
+
+A platform may choose to store the stack buildpacks and builder binary in any of the following images:
+* A companion image of the launch-image
+* A builder image that is available when rebasing
```

Comment:
We talked about this once before. IIRC the problem we are trying to solve is that a platform may wish to provide the _exact_ same stackpacks as the user used on first build for _all_ future rebases. If you always provide stackpacks in the new run image, you are upgrading the stackpacks with each rebase. This could be desirable for a tightly controlled platform, but it could be undesirable if your goal is to ensure the least amount of breakage without the user having to rebuild. The platform may wish to change the behavior of a stackpack to require new metadata from the requiring buildpacks, but this would not be a safe operation if the buildpacks are always updated. By allowing the stackpacks to be stored in a companion image, you could ensure the results of a rebase are consistent.